File: src/main/java/de/flavormate/ba_entities/unit/repository/UnitLocalizedRepository.java
/* Licensed under AGPLv3 2024 */
package de.flavormate.ba_entities.unit.repository;

import de.flavormate.ba_entities.unit.model.UnitLocalized;
import java.util.List;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

public interface UnitLocalizedRepository extends JpaRepository<UnitLocalized, Long> {
  @Query(
      "SELECT lu FROM UnitLocalized lu JOIN lu.unitRef u WHERE u.id in :ids AND lu.language ="
          + " :language")
  List<UnitLocalized> findByIdsAndLanguage(
      @Param("ids") List<Long> ids, @Param("language") String language);

  @Query(
      "SELECT lu FROM UnitLocalized lu WHERE lower(lu.labelSg) = lower(:label) OR lower(lu.labelPl)"
          + " = lower(:label) OR lower(lu.labelSgAbrv) = lower(:label) OR lower(lu.labelPlAbrv) ="
          + " lower(:label)")
  List<UnitLocalized> findByLabel(@Param("label") String label);

  @Query(
      "SELECT lu FROM UnitLocalized lu WHERE (lower(lu.labelSg) = lower(:label) OR"
          + " lower(lu.labelPl) = lower(:label) OR lower(lu.labelSgAbrv) = lower(:label) OR"
          + " lower(lu.labelPlAbrv) = lower(:label)) AND lu.language = :language")
  List<UnitLocalized> findByLabelAndLanguage(
      @Param("label") String label, @Param("language") String language);
}

File: src/main/kotlin/de/flavormate/features/unit/repositories/UnitLocalizedRepository.kt
/* Licensed under AGPLv3 2024 - 2026 */
package de.flavormate.features.unit.repositories

import de.flavormate.features.unit.daos.models.UnitLocalizedEntity
import io.quarkus.hibernate.orm.panache.kotlin.PanacheQuery
import io.quarkus.hibernate.orm.panache.kotlin.PanacheRepositoryBase
import io.quarkus.panache.common.Sort
import jakarta.enterprise.context.ApplicationScoped

@ApplicationScoped
class UnitLocalizedRepository : PanacheRepositoryBase<UnitLocalizedEntity, String> {
  fun findByLanguage(sort: Sort, language: String): PanacheQuery<UnitLocalizedEntity> {
    val params = mapOf("language" to language)

    return find(
      query = "select u from UnitLocalizedEntity u where lower(u.language) = lower(:language)",
      sort = sort,
      params = params,
    )
  }

  fun findByLabelAndLanguage(label: String, language: String): PanacheQuery<UnitLocalizedEntity> {
    val params = mapOf("label" to label, "language" to language)
    return find(
      "select u from UnitLocalizedEntity u where (lower(u.labelSg) = lower(:label) OR lower(u.labelPl) = lower(:label) OR lower(u.labelSgAbrv) = lower(:label) OR lower(u.labelPlAbrv) = lower(:label)) AND language = :language",
      params,
    )
  }
}
